resharper/resharper-unity/src/CSharp/Feature/Services/QuickFixes/MoveQuickFixes/AbstractMoveQuickFix.cs

```diff
@@ -0,0 +1,151 @@
+            
+            return null;
+        }
+
+        public virtual bool IsAvailable(IUserDataHolder cache)
+        {
+            if (MonoBehaviourScript == null || ToMove == null)
+                return false;
+            if (!MonoBehaviourScript.IsValid() || !ToMove.IsValid())
+                return false;
+            return true;
+        }
+        
+        public class MoveAction : BulbActionBase
+        {
+            [NotNull] private readonly IClassDeclaration myClassDeclaration;
+            [NotNull] private readonly ICSharpExpression myExpression;
+            [NotNull] private readonly string myMethodName;
+            [CanBeNull] private readonly string myFieldName;
+
+            public MoveAction([NotNull] IClassDeclaration classDeclaration, [NotNull] ICSharpExpression expression,[NotNull] string methodName, [CanBeNull] string fieldName = null)
+            {
+                myClassDeclaration = classDeclaration;
+                myExpression = expression;
+                myMethodName = methodName;
+                myFieldName = fieldName;
+            }
+            
+            protected override Action<ITextControl> ExecutePsiTransaction(ISolution solution, IProgressIndicator progress)
+            {
+                MonoBehaviourMoveUtil.MoveToMethodWithFieldIntroduction(myClassDeclaration, myExpression, myMethodName, myFieldName);
+                return null;
+            }
+
+            public override string Text => $"Move to {myMethodName}";
+        }
+        
+        public class MoveFromLoopAction : BulbActionBase
+        {
+            private readonly ICSharpExpression myToMove;
+            private readonly ILoopStatement myLoopStatement;
+            [CanBeNull] private readonly string myVariableName;
+
+            public MoveFromLoopAction([NotNull] ICSharpExpression toMove, [NotNull] ILoopStatement loopStatement, [CanBeNull] string variableName = null)
+            {
+                myToMove = toMove;
+                myLoopStatement = loopStatement;
+                myVariableName = variableName;
+            }
+            
+            protected override Action<ITextControl> ExecutePsiTransaction(ISolution solution, IProgressIndicator progress)
+            {
+                var anchor = myLoopStatement as ICSharpStatement;
+
+                var declaredElement = MonoBehaviourMoveUtil.GetDeclaredElementFromParentDeclaration(myToMove);
+                var baseName = myVariableName ?? MonoBehaviourMoveUtil.CreateBaseName(myToMove, declaredElement);
+                var name = NamingUtil.GetUniqueName(myToMove, baseName, NamedElementKinds.Locals, de => !de.Equals(declaredElement));
+                
+                var factory = CSharpElementFactory.GetInstance(myToMove);
+                var originMyToMove = myToMove.Copy();
```

Comment:
`.Copy()` is not the best method for moving tree nodes, try `.CopyWithResolve()` instead.
The difference between `.Copy()` and `.CopyWithResolve()` is that `.CopyWithResolve()` stores original resolve results in the copied non-physical tree which are then used by tree modification to rebind references at their new location. Generally you'd almost always prefer `.CopyWithResolve()` unless it's really just a text modification

Here are some examples where it works differently:
```cs
class D
{
  public void Source()
  {
    M(); // move this
  }

  public void M()
  {
    // to this location
    // .CopyWithResolve result: this.M();
    // .Copy result: M();
    void M() { }
  }
}
```

```cs
using System;

namespace N
{
  using static S1;

  partial class C
  {
    public void M1()
    {
      Method(); // copy this
    }
  }
}

namespace N
{
  using static S2;

  partial class C
  {
    public void M2()
    {
      // to this
      // .CopyWithResolve result: S1.Method();
    }
  }
}

static class S1
{
  public static extern void Method();
}

static class S2
{
  public static extern void Method();
}
```